fdbclient/include/fdbclient/BlobCipher.h

```diff
@@ -256,104 +261,139 @@ struct BlobCipherEncryptHeaderFlagsV1 {
 // 'encrypted buffer', compared to reading only encryptionHeader and ensuring its sanity; for instance:
 // backup-files.
 
-template <uint32_t AuthTokenSize>
-struct AesCtrWithAuthV1 {
+template <class Params>
```

Comment:
Like the idea of passing `class Params` than adding more parameters to the template if needed in the future. 

However, am not sure if following is a good idea:
1. Defining `AesCtrWithAuth` replacing version specific header, for instance: `AesCtrWithAuthV1`. 
2. Pulling header and flags version from main header into component headers. 

I initially had the same implementation and then moved away from it for the following reasons:
1. For instance, a header is defined today as follows:
` struct AV1 { int a, int b, foo y; }` 

The usage of the code today is done as follows:
`
if ( struct_version == 1) {
 // parse and extract information from AV1 struct
}
`

Now if the header in future gets involved (elements added and/or removed, we shall define
` struct A2 {int a; int x; foo f; bar b; }`

Now the usage in the code for backward compatibility will become
`
if (version == 1) {
 A1 a1 = <do necessary assignment>
} else if (version == 2) {
  A2 a2 = <do necessary assignment>
}
`

Compared to the proposed scheme it might end up looking like
`
A a; 
 if ( version == 1) {
 a.<field1> = ...
a.<field2> = ... 
} else (version == 2) {
 a.<field3> = ...
}
`

I feel code #1 seems much better to read and handle (lesser accidental failures, as compiler helps catching bugs while assignment is done etc.); it may look verbose, but, then more simplistic and given header evolution might not be that common (yes we might change header, but, not super frequently I hope).

Thoughts?

Reply:
We still have both ways to evolve the schema: (1) introduce a new algo type, and (2) evolve an existing algo type. If the new schema diverge substantially from any existing algo header, then we shall do (1). If the new schema share enough common parts with one of the existing header, then a pattern like
> if (version == 1) { A1 a1 = <do necessary assignment> } else if (version == 2) { A2 a2 = <do necessary assignment> }
is not a bad pattern, in that it allow sharing code with clarity. Does it make sense?

Reply:
We still have both ways to evolve the schema: (1) introduce a new algo type, and (2) evolve an existing algo type. If the new schema diverge substantially from any existing algo header, then we shall do (1). If the new schema share enough common parts with one of the existing header, then a pattern like

```
a.a = ...;
if (version == 1) {
    a.b = ...;
} else {
    a.x = ...;
    a.y = ...;
}
```

is not a bad pattern, in that it allow sharing code with clarity. Does it make sense?

Reply:
> in that it allow sharing code with clarity
Am not so convinced about the clarity; in some ways am worried about the `if-else` code clutter that needs to happen on per-field basis in extreme case. Otoh, explicit naming of struct with VersionNumber seems more clear and readable, especially given it is a on-disk structure, more explicit the code, easier for long term maintainability.

Reply:
In some case the sharing part could be non-trivial, say if the two version share most of the fields and differ in just one field, and for the other fields we want to have a lot of validation logic for them, we will not want to repeat those logics.

that said, with this patch, you can still introduce a `AesCtrWithAuthV2` making it one of the `algoHeader` variant and rename the exiting `AesCtrWithAuth` to `AesCtrWithAuthV1`, or maybe given then a better naming than V1/V2. Say the new type can be `AesCtrWithAuthAndSomeNewFeature`.

Reply:
> In some cases the sharing part could be non-trivial, say if the two version share most of the fields and differ in just one field, and for the other fields we want to have a lot of validation logic for them, we will not want to repeat those logics.

I think we can easily abstract it out in a common header and include the fields; also, this is even used in the current design, for instance: `BlobCipherDetails` is used in both algorithm header.

Reply:
>  you can still introduce a AesCtrWithAuthV2

Precisely my point, then we don't need versioning within `AesXXX` structs, the header algoVersion is good enough to suggest what structure needs to be (de)serialized at the runtime.

Reply:
> I think we can easily abstract it out in a common header and include the fields; also, this is even used in the current design, for instance: BlobCipherDetails is used in both algorithm header.

If you are open to composition there's another way to achieve what you want with this patch
```
struct AesCtrWithAuth {
    uint8_t version;
    union {
        AesCtrWithAuthV1 v1;
        AesCtrWithAuthV2 v2;
    }

    template <class Ar> 
    void serialize(Ar& ar) {
        serializer(ar, version);
        switch (version) {
            case 1: serializer(ar, v1); break;
            case 2: serializer(ar, v2); break;
        }
    }
};
```

Reply:
I would like to leave flexibility to evolve existing schema. Also having the algo-specific types owning their own version makes the code easier to read.

Reply:
> If you are open to composition there's another way to achieve what you want with this patch

It just feels like a syntactic sugar, clients can just create an object with name `AesCtrWithAuth`; but for all practical purposes they need to make decision based on the `version`, so the client code might not look much different than explicit struct name with versions?

Reply:
>not look much different than explicit struct name with versions

exactly. so what we can do previously we can still do it. The only differences are:
1. The algorithm headers now own their own version, for example AesCtrNoAuth is able to handle it is own interaction with CLIENT_KNOBS->ENCRYPT_HEADER_AES_CTR_NO_AUTH_VERSION.
2. Additional flexibility in evolving the existing headers without introducing a new V2 type, for which whether it is a good pattern is debatable, but I think why don't we leave the flexibility.

Reply:
also it looks pretty messy to look at the high level if the encryption header have a field
```
std::variant<AesCtrV1, AesCtrV2, AesGcmV1, AesGcmV2, AesGcmV3> algoHeader;
```
it is cleaner to see
```
std::variant<AesCtr, AesGcm> algoHeader;
```
with the versioning for each of the algo encapsulate inside each header type's own implementation.

Reply:
In reality both the version are `ugly`, but, yes for more flexibility they are what they are. Am not sure I would say one is cleaner than other, as in near future one can easily evolve it to:
`std::variant<AesCtr, AesGcm, TripleDES, RSA, BlowFish, TwoFish, RSA,....>`

So even if we have one 1 variation of the header, the list would look same. We can easily abstract it out by using a `using XXX = std::variant<xxx>` if the goal is beautification. 

Again, my concern here is version contained in the component headers, requires client code to be written with extra care, one mistake of `not` setting a required field for that particular version is a recipe for disaster. The compilation would be all fine, but, functionally it won't be. However, with explicit VersionHeader scheme, we can even define specialized classes to deal with it, plus, simple memory assignment would fail. For instance: if i try to assign AUTH_AES_CMAC to object that is AesCtrWithNoV1 it fails :).

Reply:
Btw with regard to the suggestion below:

`struct AesCtrWithAuth { uint8_t version; union { AesCtrWithAuthV1 v1; AesCtrWithAuthV2 v2; } template <class Ar> void serialize(Ar& ar) { serializer(ar, version); switch (version) { case 1: serializer(ar, v1); break; case 2: serializer(ar, v2); break; } } };`

Just to clarify I like the proposal above, however, all I meant is the client code still would have to do the same `if-else` based on the header version, but, am advocate of header evolution happening in a more explicit manner. 

All up for doing it, if you like the idea as well.

Reply:
If
```
std::variant<AesCtr, AesGcm, TripleDES, RSA, BlowFish, TwoFish, RSA,....>
```
is messy, then
```
std::variant<AesCtrV1, AesCtrV2, AesCtrV3, AesGcmV1, AesGcmV2, AesGcmV3, TripleDESV1, TripleDESV2, TripleDESV3, RSAV1, RSAV2, RSAV3, BlowFishV1, BlowFishV2, TwoFishV3, TwoFishV2, TwoFishV3, RSAV1, RSAV2, RSAV3,....>
```
is hell.. someone come here looking at `algoHeader` probably want to figure out what are the list of supported algorithms, and not for their versioning. We are deliberately mixing two layers of abstraction.

If you want I can make the change I proposed in https://github.com/apple/foundationdb/pull/9345#discussion_r1103350239, adding back the V1 header, but with it nested inside the non-versioned algorithm struct.

Reply:
I think we just crossed messages, read my reply at https://github.com/apple/foundationdb/pull/9345#discussion_r1103499174. I think we are on the same page here. Great discussion, thanks for doing this, Yi.

Reply:
> all I meant is the client code still would have to do the same if-else based on the header version

well, with that we can still enforce the "V1" client code to only deal with the V1 sub-headers and so on, achieving what you suggested.